cou/utils/openstack.py

```diff
@@ -13,125 +13,129 @@
-        "swift": ["swift-proxy", "swift-storage"],
-        "nova": ["nova-cloud-controller", "nova-compute"],
-        "ovn": ["ovn-dedicated-chassis", "ovn-central"],
-        "neutron": ["neutron-api", "neutron-gateway"],
-        "manila": ["manila-ganesha"],
-        "horizon": ["openstack-dashboard"],
-    }
-    with open("cou/utils/openstack_lookup.csv") as csv_file:
+    with open(CSV_FILE) as csv_file:
         openstack_lookup = OrderedDict()
-        service_column_index = 0
-        minor_column_index = 1
-        micro_column_index = 2
-        version_start_column_index = 3
         csv_reader = csv.reader(csv_file, delimiter=",")
         header = next(csv_reader)
         for row in csv_reader:
             service_dict: defaultdict[str, Any] = defaultdict(OrderedDict)
-            service = row[service_column_index]
-            minor = row[minor_column_index]
-            micro = row[micro_column_index]
-            for column_index in range(version_start_column_index, len(row), 2):
+            service = row[SERVICE_COLUMN_INDEX]
+            for column_index in range(VERSION_START_COLUMN_INDEX, len(row), 2):
                 os_version, _ = header[column_index].split("-")
                 min_version = row[column_index]
                 max_version = row[column_index + 1]
-                service_dict[os_version] = Versions(
+                service_dict[os_version] = VersionRange(
                     Version(min_version),
                     Version(max_version),
-                    True if minor == "TRUE" else False,
-                    True if micro == "TRUE" else False,
                 )
             openstack_lookup[service] = service_dict
-    for charm_type, charms in charms_types.items():
+    for charm_type, charms in CHARM_TYPES.items():
         for charm in charms:
             openstack_lookup[charm] = openstack_lookup[charm_type]
     return openstack_lookup
 
 
-OPENSTACK_LOOKUP = generate_openstack_lookup()
-
+class OpenStackLookup(object):
+    _OPENSTACK_LOOKUP: OrderedDict = OrderedDict()
+
+    def __new__(cls):
+        """Magic method for create the OpenStackLookup object.
+
+        :return: OpenStackLookup object
+        :rtype: OpenStackLookup
+        """
+        if not cls._OPENSTACK_LOOKUP:
+            cls._OPENSTACK_LOOKUP = generate_openstack_lookup()
+        return super().__new__(cls)
+
+    def get_compatible_openstack_codenames(
+        self, component: str, version: str
+    ) -> List[Optional[str]]:
+        """Get the possible OpenStack codenames based on the component and version.
+
+        :param component: Name of the component. E.g: "keystone"
+        :type component: str
+        :param version: Version of the component. E.g: "17.0.2"
+        :type version: str
+        :return: Return a list with the compatible OpenStack codenames.
+        :rtype: List[Optional[str]]
+        """
+        possible_os_releases: List[Optional[str]] = []
```

Comment:
In what case the `openstack_release` can be None? If there is situation like that I think we should not append it to the list.